overlord/patch/patch5.go

```diff
@@ -0,0 +1,81 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patch
+
+import (
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/wrappers"
+)
+
+func init() {
+	patches[5] = patch5
+}
+
+type log struct{}
+
+func (log) Notify(status string) {
+	logger.Noticef("patch 5: %s", status)
+}
+
+func patch5(st *state.State) error {
+	log := log{}
+
+	snapStates, err := snapstate.All(st)
+	if err != nil {
+		return err
+	}
+
+	for snapName, snapState := range snapStates {
+		if !snapState.Active {
+			continue
+		}
+
+		info, err := snapState.CurrentInfo()
+		if err != nil {
+			return err
+		}
+
+		if len(info.Apps) == 0 {
+			logger.Debugf("patch 5: skipping for %q: no apps", snapName)
+			continue
+		}
+
+		err = wrappers.StopSnapServices(info, log)
```

Comment:
Will this work correctly if the snap is already stopped?  This needs to handle the case when it crashes midway through.